content/en/docs/overview/singleton-sdk/Angular Client Introduction/data-initialization.md

```diff
@@ -2,38 +2,106 @@
 title: "Data Initialization"
 date: 2019-09-24T20:08:31+08:00
 draft: false
-weight: 15
+weight: 16
 ---
 
 
-In general, for most of products, it's good enough to load all translations and patterns one time in root module when the application starts, but for some special requirements, such like: isolated translation for feature module or using lazy module to enhance performance, please refer to [Isolated feature module](../isolated-feature-module) and [Lazy Load Module](../lazy-load-module). singleton client provides two ways to load i18n data and initialize the services in root module:
+In general, for most of the products, it's good enough to load all translations and patterns one time in root module when the application starts, but for some special requirements, such like: isolated translation for feature module or using lazy module to enhance performance, please refer to [Isolated feature module](../isolated-feature-module) and [Lazy Load Module](../lazy-load-module). singleton client provides two ways to load i18n data and initialize the services in root module:
 
 ##### **Blocking loading**
 
-Get the translations from an singleton service API and keep it ready before the application renders the page.
-
+**For preloading modules** - Create a factory function that loads i18n resource data and provide that function to the APP_INITIALIZER token. The function is executed during the application bootstrap process, and the needed data is available on startup.
+
+```
+export function initVIPConfig(vipService: VIPService, localeService: LocaleService) {
+    // Specify locale, either from browser language or user's profile.
+    const currentlocale: string  = getBrowserCultureLang();
+    localeService.init(currentlocale);
+    return () => vipService.initData(I18nConfig);
+}
+
+@NgModule({
+    ...
+    imports: [
+        ...
+        HttpClientModule,
+        VIPModule.forRoot(),
+        ...
+    ],
+    providers: [{
+        provide: APP_INITIALIZER,
+        useFactory: initVIPConfig,
+        deps: [
+            VIPService,
+            LocaleService
+        ],
+        multi: true
+    }],
+    ...
+})
+export class AppModule {}
+
+```
+
+**For lazy-loading modules** - Configure i18nDataGuard as part of the route object to load i18n resource for a lazy-loading module by reading lite vip configuration from route parameters, asynchronously load the data, and have it ready by the time the module activates and initializes.
```

Comment:
`**For lazy-loading modules** - Configure i18nDataGuard as part of the route object to load i18n resource for a lazy-loading module by reading lite Singleton configuration from route parameters, asynchronously load the data, and have it ready by the time the module activates and initializes.`

Reply:
Done!